src/mlpack/core/optimizers/sdp/lrsdp_function_impl.hpp

```diff
@@ -189,5 +187,40 @@ void AugLagrangianFunction<LRSDPFunction>::Gradient(
   gradient = 2 * s * coordinates;
 }
 
+// Template specializations for function and gradient evaluation.
+// Note that C++ does not allow partial specialization of class members,
+// so we have to go about this in a somewhat round-about way.
+template <>
+inline double AugLagrangianFunction<LRSDPFunction<SDP<arma::sp_mat>>>::Evaluate(
+    const arma::mat& coordinates) const
+{
+  return EvaluateImpl(function, coordinates, lambda, sigma);
+}
```

Comment:
I'm thinking SFINAE may be a better approach here... (I haven't tested the code below.  I think it will look something like this.)

```
template<typename MatType>
inline double AugLagrangianFunction<LRSDPFunction<SDP<MatType>>>::Evaluate(
    const arma::mat& coordinates,
    const boost::enable_if<IsMatType<MatType>::result>::result* = 0)
{
  // no more need for EvaluateImpl().
}
```

`IsMatType<T>` can be defined to be some struct where `result` is true if `T = mat` or `T = sp_mat`.  Take a look at Armadillo's `traits.hpp` file.  What do you think?